src/main/java/org/kaazing/netx/ws/internal/util/OptimisticReentrantLock.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2014, Kaazing Corporation. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.kaazing.netx.ws.internal.util;
+
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+// ### TODO: 1. happens-before relationship should be established between the thread that releases the lock and the next thread
+//              that acquires the lock.
+//           2. Evaluate using Martin's sun.misc.Unsafe based AtomicSequence implementation.
+public class OptimisticReentrantLock {
+    private final AtomicReference<Thread> reference;
+    private final AtomicInteger stamp;
+
+    public OptimisticReentrantLock() {
+        this.reference = new AtomicReference<Thread>(null);
+        this.stamp = new AtomicInteger(0);
+    }
+
+    public void lock() {
+        boolean lockObtained = false;
+
+        if (Thread.currentThread() == getOwner()) {
+            stamp.incrementAndGet();
+            return;
+        }
+
+        while (!lockObtained) {
+            if (reference.compareAndSet(null, Thread.currentThread())) {
+                stamp.set(1);
+                lockObtained = true;
+            }
+        }
```

Comment:
Sorry, I originally didn't have the method signature around it, then omitted the leading check accidentally.

You probably meant to reuse `currentThread` variable, also it might be clearer with an else clause instead of a return and also clearer to avoid going through getOwner since we're already using reference directly:
```java
public void lock() {
  Thread currentThread = Thread.currentThread();
  if (currentThread == reference.get()) {
       stamp.incrementAndGet();
  }
  else {
    while (reference.compareAndSet(null, currentThread) == false);
    stamp.set(1);
  }
}
```

One last point is whether or not to include a back-off strategy inside the while loop.